helper/schema/schema.go

```diff
@@ -1431,6 +1480,79 @@ func (m schemaMap) validateConflictingAttributes(
 	return nil
 }
 
+func removeDuplicates(elements []string) []string {
+	encountered := make(map[string]struct{}, 0)
+	result := []string{}
+
+	for v := range elements {
+		if _, ok := encountered[elements[v]]; !ok {
+			encountered[elements[v]] = struct{}{}
+			result = append(result, elements[v])
+		}
+	}
+
+	return result
+}
+
+func validateExactlyOneAttributes(
+	k string,
+	schema *Schema,
+	c *terraform.ResourceConfig) error {
+
+	if len(schema.ExactlyOneOf) == 0 {
+		return nil
+	}
+
+	allKeys := removeDuplicates(append(schema.ExactlyOneOf, k))
+	sort.Strings(allKeys)
+	specified := make([]string, 0)
+	for _, exactlyOneOfKey := range allKeys {
+		if raw, ok := c.Get(exactlyOneOfKey); ok {
+			if raw == hcl2shim.UnknownVariableValue {
+				// An unknown value might become unset (null) once known, so
```

Comment:
This sorta makes me concerned for some weird edgecase bugs. This are larger Terraform questions like when does `validate` get called and how often?

Reply:
Agreed, however this code has been around for awhile and was copy pasted to be in line with `ConflictsWith`

Reply:
Terraform Core calls `ValidateResourceChange` again during apply so we can catch things like this. It does mean that there is the potential for an error not to show up until apply time, but that is generally true for all validation.